content/advent-2018/avoid-gc-overhead-large-heaps.md

```diff
@@ -0,0 +1,376 @@
+
+	for i := 0; i < 10; i++ {
+		start := time.Now()
+		runtime.GC()
+		fmt.Printf("GC took %s\n", time.Since(start))
+	}
+
+	runtime.KeepAlive(a)
+}
+```
+
+On my 2015 MBP I get the following output.
+
+```
+GC took 4.275752421s
+GC took 1.465274593s
+GC took 652.591348ms
+GC took 648.295749ms
+GC took 574.027934ms
+GC took 560.615987ms
+GC took 555.199337ms
+GC took 1.071215002s
+GC took 544.226187ms
+GC took 545.682881ms
+```
+
+The GC takes over half a second. And why should that be surprising? I've allocated 1 billion pointers. That's actually less than a nano-second per pointer to check each pointer. Which is a pretty good speed for looking at pointers.
+
+# So what next?
+
+That seems like a fundamental problem. If our application needs a large in-memory lookup table, or if our application fundamentally is a large in-memory lookup table, then we've got a problem. If the GC insists on periodically scanning all the memory we've allocated we'll lose huge amounts of the available processing power to the GC. What can we do about this?
+
+We essentially have two choices. We either hide the memory from the GC, or make it uninteresting to the GC.
+
+## Make our memory dull
+
+How can memory be uninteresting? Well, the GC is looking for pointers. What if the type of our allocated object doesn't contain pointers? Will the GC still scan it?
+
+We can try that. In the below example we're allocating exactly the same amount of memory as before, but now our allocation has no pointer types in it. We allocate a slice of a billion 8-byte ints, again this is approximately 8GB of memory.
+
+```go
+func main() {
+	a := make([]int, 1e9)
+
+	for i := 0; i < 10; i++ {
+		start := time.Now()
+		runtime.GC()
+		fmt.Printf("GC took %s\n", time.Since(start))
+	}
+
+	runtime.KeepAlive(a)
+}
+```
+
+Again, I've run this on my 2015 MBP
+
+```
+GC took 350.941µs
+GC took 179.517µs
+GC took 169.442µs
+GC took 191.353µs
+GC took 126.585µs
+GC took 127.504µs
+GC took 111.425µs
+GC took 163.378µs
+GC took 145.257µs
+GC took 144.757µs
+```
+
+The GC is considerably more than 1000 times faster, for exactly the same amount of memory allocated. It turns out that the Go memory manager knows what types each allocation is for, and will mark allocations that do not contain pointers so that the GC does not have to scan them. If we can arrange for our in-memory tables to have no pointers, then we're on to a winner.
+
+## Keep our memory hidden
+
+The other thing we can do is hide the allocations from the GC. If we ask the OS for memory directly, the GC never finds out about it, and therefore does not scan it. Doing this is a little more involved than our previous example!
+
+Here's the equivalent of our first program where we allocate a `[]*int` with a billion (1e9) entries. This time, we use the mmap syscall to ask for the memory directly from the OS kernel. Note this only works on unix-like operating systems, but there are similar things you can do on Windows.
```

Comment:
I'd mark the syscall as `mmap` (instead of same formatting as the rest of the text - mmap)

Reply:
And if you capitalize Windows, why not Unix? :)